### ISSUE TITLE
Enhanced Navigation does not reopen broken socket connection

### ISSUE THREAD

**Issue by garrettlondon1** [dotnet/aspnetcore#55127]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

If WebSocket disconnects, then enhanced navigation goes to a page with another InteractiveServer components, the interactive component does not re-establish a connection

### Expected Behavior

If Blazor is navigating to a page with an InteractiveServer component, and there is no WebSocket connection, or the existing WebSocket connection is severed, Enhanced navigation should initialize a new WebSocket connection after navigating

### Steps To Reproduce

Create a StreamRendered component with an non-prerendered Interactive island

Disconnect from websockets

Navigate between SSR pages, which work over http
Navigate back to the InteractiveServer(prerender:false) component
- Will not appear
You can also try this with prerender:true component
- Component will appear, without interactivity

I have a private video I can share of this as well.

### .NET Version

8.0.4

**Comment by garrettlondon1** [comment]
Note: This cannot be tested with developer tools network throttling.

Wifi has to be disconnected in order to reproduce successfully.

Steps:
- Navigate to interactive page
- Turn off Wifi
- Reconnect to site
- "Websocket connected" in js console
- Interactive components still do not work

**Comment by javiercn** [comment]
@garrettlondon1 thanks for contacting us.

This seems like a bug. Probably there's some missing logic in our JS to try and re-stablish the connection if not available.

**Comment by garrettlondon1** [comment]
https://github.com/dotnet/aspnetcore/assets/3959577/ef2ba95b-6b09-4cab-874f-d865d647b109

**Comment by garrettlondon1** [comment]
I think you can only appreciate the gravity of the situation when pre-rendering is disabled.

What will happen is, your skeleton will load for interactive components, but none of your data. So your users will be switching between fully functional pages, depending on which render mode it is, and completely non-functional pages, when the render mode is InteractiveServer.

As demonstrated in the above video with prerendering true, the only effect is that the counter does not work.

In the below video, the entire counter component is not rendered, and a blank page appears. This is using the default Blazor Web template, only difference is switching Counter to @rendermode @(new InteractiveServerRenderMode(false))

https://github.com/dotnet/aspnetcore/assets/3959577/4688516a-461e-41e5-bb06-6bff3eb8dc7e

**Comment by garrettlondon1** [comment]
Have changed the title to reflect the two issues here:

1. InteractiveServer on a disconnect (as shown above), does not reopen broken socket connection on enhanced navigation

2. Enhanced Navigation **(when disabled)**, reopens a brand new circuit for the user. I would expect that it would reuse the existing circuit if not broken.

**Comment by garrettlondon1** [comment]
Another issue with InteractiveServer and NavigationManager/Anchor tags

In .NET7, when you would use NavigationManager.NavigateTo, with global interactivity.. scoped services would be not be disposed when navigating.

When using static interactivity with Interactive @pages, the opposite is the case. Scoped services are disposed on navigation, even if the websocket is not recreated.

@danroth27 am I missing anything here? I am running into huge blockers when moving so SSR

**Comment by garrettlondon1** [comment]
Another issue:

InteractiveServer components cannot catch navigation changes outside of the Interactive boundary. So editform states have no way to prevent navigation when dirty

**Comment by garrettlondon1** [comment]
Additionally,

If you **disable** enhanced navigation, why does it not reuse the existing websocket?

Currently, it creates a new SignalR connection every time you navigate

**Comment by garrettlondon1** [comment]
@javiercn with the release of .NET9 coming soon, I want to get an understanding of where this falls in priority. It's a huge bug for us. I spoke to Dan at build about possibly getting it in this release.

If the team doesn't have bandwidth would you welcome a PR for this change?

cc @danroth27

**Comment by silverlight-enjoyer** [comment]
Any news on this? Still experiencing this in the latest .net10.

**Comment by matthetherington** [comment]
We are running into this issue because we have a custom CircuitHandler that's doing some async work in OnConnectionUpAsync.

On loading the app, the user sees the server rendered content, but until the OnConnectionUpAsync Task completes any page navigation will break the circuit.

My workaround was to make that async work fire and forget, so instead of:

```csharp
public override async Task OnConnectionUpAsync(
    Circuit circuit,
    CancellationToken cancellationToken
)
{
  await MyLongRunningTask();
}
```

we have:

```csharp
public override Task OnConnectionUpAsync(
    Circuit circuit,
    CancellationToken cancellationToken
)
{
  _ = MyLongRunningTask();
  return Task.CompletedTask;
}

private async Task MyLongRunningTask()
{
  try 
  {
    await DoLongRunningThing();
  }
  catch(Exception ex)
  {
    _logger.LogError(ex, "Failed to do xyz");
  }
}
```

Would changing Blazor internally to do effectively the same thing be a viable fix? I can't envision a scenario where you'd need to block the connection from starting until something completes, the overrides that CircuitHandler makes available are usually used more like event handlers - you just want a notification when the circuit starts / connects / disconnects / ends.

There's another related issue which seems more fitting for this discussion, but it was closed as a duplicate: https://github.com/dotnet/aspnetcore/issues/63622

**Comment by garrettlondon1** [comment]
@matthetherington please see https://github.com/dotnet/aspnetcore/issues/59657 also

I believe this is a bug, but curious what the community thinks and how they are solving this outside of custom circuit handlers!

**Comment by silverlight-enjoyer** [comment]
I remember this bug causing me a ton of grief when I encountered it.

Really easy to replicate as well:

```csharp
        public override async Task OnConnectionUpAsync(
            Circuit circuit,
            CancellationToken cancellationToken)
        {
               await Task.Delay(10000);
        }
```
You can't even wrap the async work in a try catch block and catch the `TaskCancelledException` since it throws a circuit error and you get a YBOD.


In my case I was trying to preload some data for a user by fielding an async call to a database once their circuit connects. But, the most you can do is get the auth state.


```csharp
        public override async Task OnConnectionUpAsync(
            Circuit circuit,
            CancellationToken cancellationToken)
        {
               //await Task.Delay(10000);
               var authState = authStateProvider.GetAuthenticationStateAsync();
        }
```

The issue is that the handler is very sensitive to any long running async code. So, I'm kind of lucky that fetching the auth state is relatively fast.

Anyways. My solution was to instead cache the auth state and fetch the user's information from the database using the cached auth state on demand. Like so:

```csharp
public async Task<UserData> GetUserDataFromDatabase() {
     if(userData is null)
     {
        await LoadUserDataFromAuthStateCache();
     }
}

public async Task LoadUserDataFromAuthStateCache() { ... }
```

Kind of annoying and required a lot of refactoring... But works nontheless.

TL;DR: `OnConnectionUpAsync()` can't handle long async. Cache the user's auth state and use it later instead of fielding long running async calls like DB/EF commands.

**Comment by matthetherington** [comment]
Hi @garrettlondon1 

> I believe this is a bug, but curious what the community thinks and how they are solving this outside of custom circuit handlers!

Yes, I think there are two related but separate issues here:

1. Performing async work inside OnConnectionUpAsync in a custom CircuitHandler means the circuit will die if a user tries to navigate before the OnConnectionUpAsync Task completes.
2. If a websocket connection breaks, it isn't reconnecting after navigating to an InteractiveServer page 

Issue 1 wouldn't be quite as bad if the connection re-established itself afterwards, but it's an issue nonetheless.

I've left the logs I'm seeing for issue 1 over at https://github.com/dotnet/aspnetcore/issues/59657#issuecomment-5032702762

**Comment by garrettlondon1** [comment]
> Hi @garrettlondon1 
> 
> > I believe this is a bug, but curious what the community thinks and how they are solving this outside of custom circuit handlers!
> 
> Yes, I think there are two related but separate issues here:
> 
> 1. Performing async work inside OnConnectionUpAsync in a custom CircuitHandler means the circuit will die if a user tries to navigate before the OnConnectionUpAsync Task completes.
> 2. If a websocket connection breaks, it isn't reconnecting after navigating to an InteractiveServer page 
> 
> Issue 1 wouldn't be quite as bad if the connection re-established itself afterwards, but it's an issue nonetheless.
> 
> I've left the logs I'm seeing for issue 1 over at https://github.com/dotnet/aspnetcore/issues/59657#issuecomment-5032702762


You are correct about both of these! Thanks for clarifying

**Comment by vendasankarsf3945** [comment]
Hi @garrettlondon1,

I retested this on the latest .NET 11 RC version and I’m no longer able to reproduce the issue.
I also captured a short video of the current behavior, attached below. In the recording, I:

- opened the app,
- navigated between the SSR page and the interactive Counter page,
- broke and restored the network/WebSocket connection,
- and then revisited the interactive page.

In the latest version, the interactive page reconnects and continues working as expected after navigation. I do not see the blank or non-interactive state that was previously reported.
This appears to be fixed in the latest .NET 11 RC version I tested.
If you are still able to reproduce the problem on the latest .NET version, please share a minimal repro sample along with detailed reproduction steps so we can investigate further.

<img width="1920" height="922" alt="Image" src="https://github.com/user-attachments/assets/0169cceb-9962-4098-a589-2037d0ab3233" />
